generators/typescript/codegen/src/ast/core/Writer.ts

```diff
@@ -1,7 +1,149 @@
+            if (existing == null) {
+                this.defaultImports[reference.importFrom.moduleName] = reference;
+            } else if (existing.name !== reference.name) {
+                throw new Error(
+                    `Cannot have multiple default imports for module ${reference.importFrom.moduleName}: ` +
+                        `got ${reference.name} but already had ${existing.name}`
+                );
+            }
+        }
+    }
+
+    private addNamedImport(reference: Reference): void {
+        if (reference.importFrom?.type === "named") {
+            const existing = this.imports[reference.importFrom.moduleName];
+            if (existing != null) {
+                const existingStar = existing.filter((e) => e.importFrom?.type === "star");
+                if (existingStar.length > 0) {
+                    throw new Error(
+                        `Cannot add named import ${reference.name} because non-named` +
+                            ` imports ${existingStar.map((e) => e.name)} already exist`
+                    );
+                }
+            }
+        }
+    }
+
+    private addStarImport(reference: Reference): void {
+        if (reference.importFrom?.type === "star") {
+            const existing = this.imports[reference.importFrom.moduleName];
+            if (existing != null) {
+                const existingNamed = existing.filter((e) => e.importFrom?.type === "named");
+                if (existingNamed.length > 0) {
+                    throw new Error(
+                        `Cannot add non-named import ${reference.name} because named` +
+                            ` imports ${existingNamed.map((e) => e.name)} already exist`
+                    );
+                }
+            }
+            const moduleForAlias = this.starImportAliasesInverse[reference.importFrom.starImportAlias];
+            if (moduleForAlias != null && moduleForAlias !== reference.importFrom.moduleName) {
+                throw new Error(
+                    `Attempted to use alias ${reference.importFrom.starImportAlias} for more than one ` +
+                        "module in the same file"
+                );
+            }
+            const existingAlias = this.starImportAliases[reference.importFrom.moduleName];
+            if (existingAlias == null) {
+                this.starImportAliases[reference.importFrom.moduleName] = reference.importFrom.starImportAlias;
+                this.starImportAliasesInverse[reference.importFrom.starImportAlias] = reference.importFrom.moduleName;
+            } else if (existingAlias != null && existingAlias !== reference.importFrom.starImportAlias) {
+                throw new Error(
+                    "Cannot have more than one alias for non-named imports from a module: " +
+                        `got ${reference.importFrom.starImportAlias} but already have ${existingAlias}.`
+                );
+            }
+        }
+    }
```

Comment:
There's a fair amount of nesting in these functions - I'd try to reduce nesting as much as possible and split these out into separate functions so that it's a little easier to read and debug later on.

I think we can apply an approach similar to the one outlined in [this comment](https://github.com/fern-api/fern/pull/5066#discussion_r1826146433). For example, reversing the `if` condition at the top would reduce the entire function nesting:

```typescript
if (reference.importFrom?.type !== "star") {
  return;
}
...
```

In summary, I'd try to:
1. Reduce nesting and return early as much as possible. Although [this guide](https://github.com/uber-go/guide/blob/master/style.md#reduce-nesting) is focused on Go, the same ideas apply here.
2. Add `*OrThrow` functions to handle error cases.
3. With the restructured control flow, simplify the conditions used in each `if` down to a minimum.